### DISCUSSION TITLE
Convert 16bit transparent image to 8bit transparent image, using dither

### DISCUSSION THREAD

**Discussion by kalamazandy** [ImageMagick/ImageMagick#5255] [Help]
Some 3D rendering methods don't do a great job of dithering values so things like ground shadows, especially when transparent, cause banding. At 16bit, the banding is effectively eliminated, so I wondered if I could render to 16 bit and convert to 8. But I'm failing at even getting transparency to work. I'm not sure how the dithering works, and if it even works on the alpha values, or only for RGB? I recently had to convert some renders to RGB565 and ImageMagick worked Amazingly well.

The code I'm trying right now is:
`convert test16.png -quantize Log -dither FloydSteinberg -remap allColors.webp -define png:format=png24 test16_dither.png`

I'm remapping the colors to force the image to remap to an image that has every single RGB888 color in it. Although, perhaps that is what is messing some things up, because that doesn't have values of transparency?
There are So many options that I'm not sure what I can/should use. A coworker was using `threshold,32,64,32` to change to RGB565, but I'm not sure if that was working correctly or if it's just like using posterize on an image, limiting to 32 levels. Because what I really want to do is limit to the colors of RGB565 in that case, or in this case I want to get RGBA 8888 (not sure what that format would actually be named).

The goal here is to take the 16 bit color image and dither it using FloydSteinberg so that the more limited 8bit/channel image would look close to the 16 bit color image, but without banding in smooth transition areas. Those that don't do renderings would most frequently see those issues in pictures with sky.

**Comment by snibgo** [comment]
> But I'm failing at even getting transparency to work.
... -define png:format=png24 ...

PNG24 format is three channels (R, G, B) at 8 bits/channel. So it doesn't record transparency, which might explain the problem with "getting transparency to work". Try PNG32 format.

Don't forget that when processing an image (such as colour reduction), you can save transparency, make the image opaque, do the processing, and restore transparency.

> The goal here is to take the 16 bit color image and dither it using FloydSteinberg so that the more limited 8bit/channel image would look close to the 16 bit color image, but without banding in smooth transition areas.

Can you post a sample 16-bit input?

**Reply by kalamazandy** [reply]
Oh wow. Well, not I'm embarrassed at that stupid mistake. I blame Adobe. lol
Photoshop lists PNG8 as 256 colors and PNG24 as 8 bit color and just has a checkbox to allow/not transparency. 32bit is generally reserved for per color definitions, so that would be floating point. And I am aware that PNG24 is 8 bits per channel, so I Definitely should have been able to guess that PNG32 was Also 8 bits per channel, but include alpha, because.....that's easy math!

I had thought about making sure the dither happened without the transparency, but realistically it shouldn't have to because the shadow would be the transparent part here, and the color Should be black, where the alpha channel would be the part that needs transparency. Of course, this being a render from Keyshot, that isn't actually true, so that's a good reason to do it that way. I don't know how to do that, so will have to do some research. 

This is my test image. When converting from 16bit to 8bit color you'll see harsh banding in the shadow. It's not a particularly great image, I was just trying to create an example where I actually got that. Tif wasn't supported, and it's 7MB, so I just shared it as oneDrive. 
[https://teknainc-my.sharepoint.com/:i:/g/personal/aengelkemier_teknateam_com/EewXdR026K9AmreGmHfXH2MBEHDIZuLWAksQpPBCPWzKAQ?e=f4rTxh](url)

To process just the RGB as one image and the Alpha of another, then combine them again, can you do that just in memory, or do you have to do that with temporary images? I'm doing it all in a batch file (I'm on windows) and I've made a file that looks at all of the images in a folder, running the same settings on all of the files in that folder. That way I don't have to write new code, I just drop the files in that folder, run it, and move them out when it's done.

**Comment by snibgo** [comment]
> To process just the RGB as one image and the Alpha of another, then combine them again, can you do that just in memory, or do you have to do that with temporary images? 

You can do it in a single command, so temporary files are not needed. For example (Windows syntax):
```
magick file.tif[0] ^
  ( +clone -alpha extract -write mpr:ALPH +delete ) ^
  -alpha off ^
  ... your_processing ... ^
  mpr:ALPH -compose CopyOpacity -composite ^
  out.tif
```
Use any name you like instead of "ALPH".
